### ISSUE TITLE
separate regexes from lisp file?

### ISSUE THREAD

**Issue by bounceme** [jacktasia/dumb-jump#136]
it seems like this could be of use, with a few modifications, outside of emacs, as a lot of the logic is external command related. could this work with the current architecture of this emacs mode?

**Comment by jacktasia** [comment]
Yes, I have often thought about just storing the regexes in json so that I could make dumb-jump for other text-editors and they could all just share the same json file. Unfortunately, I don't have the time to do that right now but would definitely consider a PR to do this.

**Comment by bounceme** [comment]
thanks, i'll be looking into this more when i get the time

**Comment by jacktasia** [comment]
> thanks, i'll be looking into this more when i get the time

Cool, if/when you do please discuss the json structure and overall plan before actually doing all the work on the PR. Just to ensure we all stay on the same page and you don't waste any time. Thanks!

**Comment by bounceme** [comment]
https://gist.github.com/bounceme/6e36a93faaa6d0885eded3e6560f4ddc

 here is some some hopefully correct data

**Comment by netromdk** [comment]
That's nice but I think @jacktasia wanted it to be JSON so it is easy to load from all external tools and variations.

**Comment by bounceme** [comment]
this isn't the pr. i'll work on a pr later, first need to figure out how to parse json to the equivalent elisp data

**Comment by bounceme** [comment]
https://gist.github.com/bounceme/ae8bab4f56ec1948ac1faca3b2f38b98

this is possibly incorrect, i did this manually with regex, then prettified it

**Comment by bounceme** [comment]
if someone wants to volunteer to get this working, since locally compiling and editing elisp is really hard for a vim user, it should _similar_ to this(i think):
```lisp
(defcustom dumb-jump-find-rules
  (json-read-file (expand-file-name "jump-extern-defs.json" (file-name-directory load-file-name)))
;; changed part^^
  "List of regex patttern templates organized by language and type to use for generating the grep command."
  :group 'dumb-jump
  :type
  '(repeat
    (plist
     :options ((:type string)
               (:supports string)
               (:language string)
               (:regex string)
               (:tests (repeat string))
               (:not (repeat string))))))
```

**Comment by bounceme** [comment]
another thing to know is that some json decoders don't like the duplicate key names e.g. "type"

**Comment by jacktasia** [comment]
@bounceme  Thanks for working on this.

> another thing to know is that some json decoders don't like the duplicate key names e.g. "type"

Makes sense. The format [here](https://gist.github.com/bounceme/ae8bab4f56ec1948ac1faca3b2f38b98) is definitely wrong...it's one giant json object (which is the issue with duplicate keys). 

I think the ideal way would be a list/array of objects like this (example of first two rule sets):

```
[
  {
    "type": "function",
    "supports": [
      "ag",
      "grep",
      "rg",
      "git-grep"
    ],
    "language": "elisp",
    "regex": "((defun|cl-defun)\\s+JJJ\\j",
    "tests": [
      "(defun test (blah)",
      "(defun test\n",
      "(cl-defun test (blah)",
      "(cl-defun test\n"
    ],
    "not": [
      "(defun test-asdf (blah)",
      "(defun test-blah\n",
      "(cl-defun test-asdf (blah)",
      "(cl-defun test-blah\n"
    ]
  },
  {
    "type": "variable",
    "supports": [
      "ag",
      "grep",
      "rg",
      "git-grep"
    ],
    "language": "elisp",
    "regex": "\\(defvar\\b\\s*JJJ\\j",
    "tests": [
      "(defvar test ",
      "(defvar test\n"
    ]
  }
]
```

**Comment by bounceme** [comment]
gist updated

**Comment by jacktasia** [comment]
> gist updated

Awesome, that looks good!

**Comment by bounceme** [comment]
https://github.com/jacktasia/dumb-jump/blob/master/dumb-jump.el#L322 is that intentional?

**Comment by jacktasia** [comment]
>  is that intentional?

Nope. The second `:tests` was supposed to be a `:not` but since it didn't override it (it would fail) it just got lost. It's fixed now. Thanks for the heads up!

**Comment by bounceme** [comment]
https://github.com/bounceme/dim-jump is shaping up quite well. proving useful already! don't want this too seem impatient, just a proof of the benefits

**Comment by jacktasia** [comment]
> https://github.com/bounceme/dim-jump is shaping up quite well. proving useful already! don't want this too seem impatient, just a proof of the benefit to use json

That's cool! The main thing that is blocking this for me is that I need to sync this change with MELPA because the current receipe it uses wouldn't auto include a `.json` file. That said, I am flirting with the idea of dumb-jump pulling down the json from github automatically if it's not found relatively.

Have you thought about how you're going to keep the json synced on your end?

**Comment by bounceme** [comment]
I think i'll just make the json file a user provided thing. with a link to it, it's only a cURL away. and if someone wants to contribute regexes they can all go to one place
